### PR TITLE
Forward failed storage-api row inserts to the failedStorageApiInserts PCollection addresses #23628

### PR DESCRIPTION
**Pull request by reuvenlax**
Detect row-level errors returned by AppendRows, forward them to the failed-rows PCollection, retrying remaining rows. Size limits are limits on the request size itself, so also detect those large rows separately and forward them to the failed-rows collection.

### CODE REVIEW

**Comment by reuvenlax** [comment]
R: @yirutang

**Comment by reuvenlax** [comment]
Run Spotless PreCommit

**Comment by reuvenlax** [comment]
run PostCommit Java DataflowV1

**Comment by reuvenlax** [comment]
run DataflowV1 Java PostCommit

**Comment by reuvenlax** [comment]
Run PostCommit_Java_DataflowV1

**Comment by reuvenlax** [comment]
Run PostCommit_Java_Dataflow

**Comment by reuvenlax** [comment]
Run Python PreCommit

**Comment by reuvenlax** [comment]
Run Python PreCommit

**Comment by reuvenlax** [comment]
friendly ping!

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 1]
```diff
@@ -328,29 +375,62 @@ void addMessage(StorageApiWritePayload payload) throws Exception {
...
+        if (inserts.getSerializedSize() >= maxRequestSize) {
+          if (inserts.getSerializedRowsCount() > 1) {
+            // TODO(reuvenlax): Is it worth trying to handle this case by splitting the protoRows?
+            // Given that we split
+            // the ProtoRows iterable at 2MB and the max request size is 10MB, this scenario seems
+            // nearly impossible.
+            LOG.error(
+                "A request containing more than one row is over the request size limit of "
+                    + maxRequestSize
+                    + ". This is unexpected. All rows in the request will be sent to the failed-rows PCollection.");
```
Comment (chamikaramj) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:401]: Should we still try to send it once ?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java [hunk 2]
```diff
@@ -415,130 +442,217 @@ public String toString() {
...
-        context.numRows = protoRows.getSerializedRowsCount();
-        contexts.add(context);
-        retryManager.addOperation(run, onError, onSuccess, context);
-        recordsAppended.inc(protoRows.getSerializedRowsCount());
-        appendSizeDistribution.update(context.numRows);
-      }
-      initializeContexts.accept(contexts, false);
+            // Invalidate the StreamWriter and force a new one to be created.
+            LOG.error(
+                "Got error " + failedContext.getError() + " closing " + failedContext.streamName);
```
Comment (chamikaramj) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:524]: The logic here is long and hard to follow. I wonder if we can better structure this someone for anybody in the future who is trying to get an abstract idea regarding the logic here (and may be modify it).

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java_GCP_IO_Direct PreCommit

**Review by chamikaramj — COMMENTED** [review]
Thanks. LGTM.

**Comment by reuvenlax** [comment]
Run Python PreCommit

**Comment by reuvenlax** [comment]
Run PostCommit_Java_Dataflow

**Comment by yirutang** [comment]
R:@agrawal-siddharth

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 3]
```diff
@@ -328,29 +375,62 @@ void addMessage(StorageApiWritePayload payload) throws Exception {
...
-        inserts.addAllSerializedRows(pendingMessages);
 
-        ProtoRows protoRows = inserts.build();
+        final ProtoRows.Builder insertsBuilder = ProtoRows.newBuilder();
+        insertsBuilder.addAllSerializedRows(pendingMessages);
+        final ProtoRows inserts = insertsBuilder.build();
         pendingMessages.clear();
 
+        // Handle the case where the request is too large.
+        if (inserts.getSerializedSize() >= maxRequestSize) {
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:392]: Unfortunately it still is.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java [hunk 4]
```diff
@@ -415,130 +442,217 @@ public String toString() {
...
+            // duplicates.
+            boolean offsetMismatch =
+                statusCode.equals(Code.OUT_OF_RANGE) || statusCode.equals(Code.ALREADY_EXISTS);
+            // This implies that the stream doesn't exist or has already been finalized. In this
+            // case we have no choice but to create a new stream.
+            boolean streamDoesNotExist =
+                explicitStreamFinalized
+                    || statusCode.equals(Code.INVALID_ARGUMENT)
+                    || statusCode.equals(Code.NOT_FOUND)
+                    || statusCode.equals(Code.FAILED_PRECONDITION);
```
Comment (yirutang) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:547]: Why FAILED_PRECONDITION belongs to streamDoesNotExist?

Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:547]: I don't 100% remember, but I believed FAILED_PRECONDITION was being returned when we tried to write to a stream that had been finalized.

**Comment by reuvenlax** [comment]
Run PostCommit_Java_Dataflow

**Comment by reuvenlax** [comment]
Run Spotless PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by reuvenlax** [comment]
Run Java_Kafka_IO_Direct PreCommit

**Comment by reuvenlax** [comment]
Run Java_Neo4j_IO_Direct PreCommit

**Comment by reuvenlax** [comment]
Run Python PreCommit

**Comment by reuvenlax** [comment]
Run SQL PreCommit

**Comment by reuvenlax** [comment]
Run Java_Neo4j_IO_Direct PreCommit

**Comment by reuvenlax** [comment]
Run Python PreCommit

**Comment by reuvenlax** [comment]
Run Java_Neo4j_IO_Direct PreCommit

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java (hunk 2 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:524]: ack - worth thinking if we can structure it differently.
